### PR TITLE
ddl: add foreign key compatibility for temporary table

### PR DESCRIPTION
**Pull request by Howie59**
### What problem does this PR solve?

Issue Number: see https://github.com/pingcap/tidb/pull/24168

### What is changed and how it works?

Proposal: [temporary table](https://github.com/pingcap/tidb/issues/24169) 

What's Changed:  add foreign key compatibility for temporary table

### Check List 

Tests 

- Unit test

### Release note 

- No release note

### CODE REVIEW

File: ddl/ddl_api.go [hunk 1]
```diff
@@ -5264,6 +5264,9 @@ func (d *ddl) CreateForeignKey(ctx sessionctx.Context, ti ast.Ident, fkName mode
 	if err != nil {
 		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ti.Schema, ti.Name))
 	}
+	if t.Meta().TempTableType != model.TempTableNone {
+		return ErrOptOnTemporaryTable.GenWithStackByArgs("foreign key")
```
Comment (djshow832) [inline ddl/ddl_api.go:5268]: Please be compatible with MySQL.
`ERROR 1215 (HY000): Cannot add foreign key constraint`

File: ddl/ddl_api.go [hunk 2]
```diff
@@ -5296,6 +5299,9 @@ func (d *ddl) DropForeignKey(ctx sessionctx.Context, ti ast.Ident, fkName model.
 	if err != nil {
 		return errors.Trace(infoschema.ErrTableNotExists.GenWithStackByArgs(ti.Schema, ti.Name))
 	}
+	if t.Meta().TempTableType != model.TempTableNone {
```
Comment (djshow832) [inline ddl/ddl_api.go:5302]: Why check it in `DropForeignKey`? Is there any case in which we should check it?

Comment (Howie59) [reply ddl/ddl_api.go:5302]: alter table xx drop foreign key ...

Comment (Howie59) [reply ddl/ddl_api.go:5302]: don't we need to check it?

Comment (djshow832) [reply ddl/ddl_api.go:5302]: Now that temporary tables won't have foreign keys, dropping foreign keys from these tables will report errors.

File: planner/core/planbuilder.go [hunk 3]
```diff
@@ -3474,8 +3474,15 @@ func (b *PlanBuilder) buildDDL(ctx context.Context, node ast.DDLNode) (Plan, err
...
-			return nil, ErrOptOnTemporaryTable.GenWithStackByArgs("create table like")
+		if v.TemporaryKeyword != ast.TemporaryNone {
+			for _, cons := range v.Constraints {
+				if cons.Tp == ast.ConstraintForeignKey {
+					return nil, infoschema.ErrCannotAddForeign
+				}
+			}
+			if v.ReferTable != nil {
+				return nil, ErrOptOnTemporaryTable.GenWithStackByArgs("create table like")
+			}
```
Comment (djshow832) [inline planner/core/planbuilder.go:3485]: Remove this. We're capable of supporting it. Please file another PR to solve `create table like` issue.

**Review by djshow832 — APPROVED** [review]

**Pull request merged**